### PR TITLE
Added Generic Separate Boxes for Entering OTP 

### PR DESCRIPTION
**Pull request by ankita1010**
## This PR contains the following changes -

- Added OTPBox component which is reusable and the length of OTP can be modified anytime without any changes required in the OTPBox component.

- Built the popular OTP boxes containing separate input boxes for each digit.

Features: 
 - Each box accepts only one digit numbers
 - On entering a number, the focus is shifted to the next box
 - On backspace keypress, the focus is moved to the previous box (if exists)
 - Styled the boxes with a little creative freedom, but adhering to the theme

### Refer screenshot

Screenshot 2020-06-09 at 4 25 16 AM

<em>Comments and suggestions are welcome</em>.
Cheers! 🍺

### CODE REVIEW

File: src/components/otpBox/index.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    const focusToNextBox = (index, value) => {
+        parseInt(index) < lengthOfOTP - 1 && isOneDigitNumber(value) && inputsRef[parseInt(index) + 1].focus();
+    };
+
+    const focusToPreviousBox = (index) => {
+        parseInt(index) > 0 && inputsRef[parseInt(index) - 1].focus();
+    };
+
+
```
Comment (himayand21) [inline src/components/otpBox/index.js:45]: Extra lines of spaces seems unnecessary.
One line separator can be used instead.

File: src/pages/welcome/VerifyOTP.js [hunk 2]
```diff
@@ -1,23 +1,34 @@
-import React, {useState, useEffect} from 'react';
+import React, { useState, useEffect, useRef, createRef } from 'react';
```
Comment (himayand21) [inline src/pages/welcome/VerifyOTP.js:1]: createRef is not being used.
Let's not import it then.

File: src/pages/welcome/VerifyOTP.js [hunk 3]
```diff
@@ -1,23 +1,34 @@
...
 import PropTypes from 'prop-types';
-import {useHistory} from 'react-router-dom';
-import {NavBar} from '../../components/navBar';
-import {Footer} from '../../components/footer';
-import {Loader} from '../../components/loader';
-import {WELCOME} from '../../constants';
-import {Background} from '../../components/background';
-import {useToast} from '../../components/toast';
+import { useHistory } from 'react-router-dom';
```
Comment (himayand21) [inline src/pages/welcome/VerifyOTP.js:3]: Linting issue.
Check eslint throughout the file as well.

File: src/components/otpBox/index.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        else {
+            const newValue = value.slice(otp[otpid].length) || value;
+            setOtp({ ...otp, [otpid]: isOneDigitNumber(newValue) ? newValue : otp[otpid] });
+            focusToNextBox(index, value);
+        }
+    };
+
+    const checkBackSpace = (event) => {
+        const { dataset: { otpid, index, } } = event.target;
+        if ((event.key == 'Backspace' || event.keyCode === 8) && !otp[otpid]) {
```
Comment (himayand21) [inline src/components/otpBox/index.js:29]: Any specific reason for using ==, and not ===?

File: src/components/otpBox/index.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        }
+        else {
+            const newValue = value.slice(otp[otpid].length) || value;
+            setOtp({ ...otp, [otpid]: isOneDigitNumber(newValue) ? newValue : otp[otpid] });
+            focusToNextBox(index, value);
+        }
+    };
+
+    const checkBackSpace = (event) => {
+        const { dataset: { otpid, index, } } = event.target;
```
Comment (himayand21) [inline src/components/otpBox/index.js:28]: One extra comma in the end can be removed.

File: src/components/otpBox/index.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        otp,
+        setOtp,
+        inputsRef
+    } = props;
+    const lengthOfOTP = OTPInputIds.length;
+
+    const handleOTPinput = (event) => {
+        const { value, dataset: { otpid, index } } = event.target;
+        if (otp[otpid] === '') {
+            setOtp({ ...otp, [otpid]: isOneDigitNumber(value) ? value : otp[otpid] });
```
Comment (himayand21) [inline src/components/otpBox/index.js:14]: reassigning the existing state value again is not necessary here.
Instead, the isOneDigitNumber(value) can be checked before and state can be updated in that if case.
If it returns false, no need to update the state.

File: src/components/otpBox/index.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        if ((event.key == 'Backspace' || event.keyCode === 8) && !otp[otpid]) {
+            event.preventDefault();
+            focusToPreviousBox(index);
+        }
+    };
+
+    const isOneDigitNumber = (value) => !isNaN(value) && value.length === 1;
+
+    const focusToNextBox = (index, value) => {
+        parseInt(index) < lengthOfOTP - 1 && isOneDigitNumber(value) && inputsRef[parseInt(index) + 1].focus();
```
Comment (himayand21) [inline src/components/otpBox/index.js:38]: use parenthesis to wrap the comparative blocks in such cases to improve code readability.

File: src/pages/welcome/VerifyOTP.js [hunk 8]
```diff
@@ -37,10 +50,15 @@ const VerifyOTP = (props) => {
         if (user?.email) setEmail(user.email);
     }, [user]);
 
+
+    const OTPEntered = Object.values(otp).reduce((acc, currentVal) => String(acc) + String(currentVal), '');
```
Comment (himayand21) [inline src/pages/welcome/VerifyOTP.js:54]: I guess, Array.join() would have been simpler?

File: src/pages/welcome/VerifyOTP.js [hunk 9]
```diff
@@ -114,40 +132,40 @@ const VerifyOTP = (props) => {
                                             Resend
                                         </span>
                                     </div>
-                                    <button className="standard-button" onClick={handleVerifyOTP}>
+                                    <button className="standard-button" disabled={!isOTPValid()} onClick={handleVerifyOTP}>
```
Comment (himayand21) [inline src/pages/welcome/VerifyOTP.js:135]: instead of a function, let's make isOTPValid a variable.

File: src/components/otpBox/index.js (hunk 1 above)
Comment (ankita1010) [reply src/components/otpBox/index.js:45]: I did not have eslint. Apologies.

File: src/components/otpBox/index.js (hunk 4 above)
Comment (ankita1010) [reply src/components/otpBox/index.js:29]: A mistake. Will change and push the code.

**Pull request merged**